swift/extractor/SwiftOutputRewrite.cpp

```diff
@@ -0,0 +1,318 @@
+      // We don't patch the map in place as we need to collect all the input files first
+      outputFileMapIndexes.push_back(++i);
+    } else if (CLIArgs[i] == "-target") {
+      targetTriple = CLIArgs[++i];
+    } else if (CLIArgs[i][0] != '-') {
+      // TODO: add support for input file lists?
+      // We need to collect input file names to later use them to extract information from the
+      // output file maps.
+      maybeInput.push_back(CLIArgs[i]);
+    }
+  }
+
+  for (auto index : outputFileMapIndexes) {
+    auto oldPath = CLIArgs[index];
+    auto maybeNewPath = rewriteOutputFileMap(config, oldPath, maybeInput, remapping);
+    if (maybeNewPath) {
+      auto newPath = maybeNewPath.value();
+      CLIArgs[index] = newPath;
+      remapping[oldPath] = newPath;
+    }
+  }
+
+  // This doesn't really belong here, but we've got Xcode...
+  for (auto& [oldPath, newPath] : remapping) {
+    llvm::StringRef path(oldPath);
+    auto aliases = computeModuleAliases(path, targetTriple);
+    for (auto& alias : aliases) {
+      remapping[alias] = newPath;
+    }
+  }
+
+  return remapping;
+}
+
+void ensureNewPathsExist(const std::unordered_map<std::string, std::string>& remapping) {
+  for (auto& [_, newPath] : remapping) {
+    llvm::SmallString<PATH_MAX> filepath(newPath);
+    llvm::StringRef parent = llvm::sys::path::parent_path(filepath);
+    if (std::error_code ec = llvm::sys::fs::create_directories(parent)) {
+      std::cerr << "Cannot create redirected directory: " << ec.message() << "\n";
+    }
+  }
+}
+
+void storeRemappingForVFS(const SwiftExtractorConfiguration& config,
+                          const std::unordered_map<std::string, std::string>& remapping) {
+  // Only create remapping for the .swiftmodule files
+  std::unordered_map<std::string, std::string> modules;
+  for (auto& [oldPath, newPath] : remapping) {
+    if (llvm::StringRef(oldPath).endswith(".swiftmodule")) {
+      modules[oldPath] = newPath;
+    }
+  }
+
+  if (modules.empty()) {
+    return;
+  }
+
+  if (std::error_code ec = llvm::sys::fs::create_directories(config.tempVFSDir)) {
+    std::cerr << "Cannot create temp VFS directory: " << ec.message() << "\n";
+    return;
+  }
+
+  if (std::error_code ec = llvm::sys::fs::create_directories(config.VFSDir)) {
+    std::cerr << "Cannot create VFS directory: " << ec.message() << "\n";
+    return;
+  }
+
+  // Constructing the VFS yaml file in a temp folder so that the other process doesn't read it
+  // while it is not complete
+  // TODO: Pick a more robust way to not collide with files from other processes
+  auto tempVfsPath = config.tempVFSDir + '/' + std::to_string(getpid()) + "-vfs.yaml";
+  std::error_code ec;
+  llvm::raw_fd_ostream fd(tempVfsPath, ec, llvm::sys::fs::OF_None);
+  if (ec) {
+    std::cerr << "Cannot create temp VFS file: '" << tempVfsPath << "': " << ec.message() << "\n";
+    return;
+  }
+  // TODO: there must be a better API than this
```

Comment:
Somewhere in the closed source XCode codebase?

Reply:
In general - yes 😆.
But I meant some LLVM's `YAMLTraits` things. It's just felt easier to construct this by hand than to dig through the APIs.